unidist/cli/backends/mpi/runner.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) 2021 Modin authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""``MPIRunner`` class functionality."""
+
+import os
+import warnings
+
+from unidist.cli.base.runner import BackendRunner
+from unidist.cli.base.utils import (
+    Defaults,
+    BackendName,
+    validate_num_cpus,
+    validate_hosts,
+    get_localhost_ip,
+    get_unidist_root,
+)
+
+
+class MPIRunner(BackendRunner):
+    """
+    An implementation of unidist ``BackendRunner`` for MPI backend.
+
+    Parameters
+    ----------
+    **cli_kwargs : dict
+        Keyword arguments supported by unidist CLI.
+    """
+
+    def __init__(self, **cli_kwargs):
+        self.backend = BackendName.MPI
+        super().__init__(**cli_kwargs)
+
+    def check_kwargs_support(self, **kwargs):
+        """Check support for `kwargs` combination for MPI backend."""
+        hosts = kwargs.get("hosts", Defaults.HOSTS)
+        num_cpus = kwargs.get("num_cpus", Defaults.NUM_CPUS)
+
+        if not isinstance(num_cpus, list):
+            num_cpus = [num_cpus]
+
+        if not isinstance(hosts, list):
+            hosts = [hosts]
+
+        if len(hosts) != len(num_cpus):
+            # If `num_cpus` isn't provided or a single value `default` is provided
+            # all workers will use `default` value
+            if len(num_cpus) == 1 and num_cpus[0] == Defaults.NUM_CPUS:
+                num_cpus *= len(hosts)
+            else:
+                raise RuntimeError(
+                    "`num_cpus` and `hosts` parameters must have the equal number of values."
+                )
+
+        if (
+            kwargs.get("redis_password", Defaults.REDIS_PASSWORD)
+            != Defaults.REDIS_PASSWORD
+        ):
+            warnings.warn(
+                f"`redis_password` isn't supported for {self.backend} backend, ignored.",
+                RuntimeWarning,
+            )
+
+        self.hosts = validate_hosts(hosts)
+        self.num_cpus = validate_num_cpus(num_cpus)
+
+    def prepare_env(self):
+        """Setup unidist environment variables for MPI backend."""
+        super().prepare_env()
+        os.environ["UNIDIST_CPUS"] = str(sum([int(val) for val in self.num_cpus]))
+
+    def get_command(self):
+        """
+        Get a command to be run in a subprocess.
+
+        Returns
+        -------
+        list
+            List of strings with the command representation.
+        """
+        unidist_root = get_unidist_root()
+        workers_dir = "/tmp"
+        command = ["mpiexec", "-hosts"]
+        hosts_str = f"{get_localhost_ip()}:1,{get_localhost_ip()}:1,"
```

Comment:
Elaborate this, please. What would be a command for running MPI with -num_cpus 1 2 -host 1.1.1.1 1.2.3.1?

Reply:
For your example command will be:
```
mpiexec -hosts localhost_ip:1,localhost_ip:1,1.1.1.1:1,1.2.3.1:2 -n 1 python user_script.py : -n 1 -wdir /tmp python path_to_monitor.py : -n 1 -wdir /tmp python path_to_worker.py : -n 2 -wdir /tmp python path_to_worker.py
```

`hosts_str` represents pairs of `host_i:num_workers_i` for the each program in MPMD mode. In your example we have 4 hosts (2 localhost and your 2 IPs) and 4 programs (user_script.py, monitor.py, worker.py, worker.py)

Reply:
Thanks! Did you run it in a cluster?

Reply:
Yes, I've run this in the cluster. `mpiexec` runs an expected number of workers on each node.

Reply:
Ok, cool!